### DISCUSSION TITLE
Win exe - Can these parts be further compressed?

### DISCUSSION THREAD

**Discussion by necros2k7** [upx/upx#657] [Ideas]
This issue tracker is **ONLY** used for requesting sensible features.
Please use [stackoverflow](https://stackoverflow.com) for supporting issues.

## What's the problem (or question)?

Can empty space in windows executables be further compressed?
start of exe
![start](https://user-images.githubusercontent.com/16452453/224519245-092fc1c6-3913-4cfc-b37b-238fff5f30a0.png)
end of exe
![end](https://user-images.githubusercontent.com/16452453/224519253-cd7a6935-f4d4-48f8-91b3-c6b17bf4e859.PNG)
## What should have happened?

Compressing these parts will further reduce file size
## How can we reproduce the issue?

1. Compress any windows binary
2.
3.
4.

## Please tell us details about your environment.

* UPX version used (`upx --version`): latest actions build
* Host Operating System and version: win11
* Host CPU architecture: x64
* Target Operating System and version: win11
* Target CPU architecture: x64

**Comment by jreiser** [comment]
You must upload ("Attach files by ...") or give a URL to an actual PE .exe file, and then describe the regions which you suspect might be further compressed, naming  ranges of numerical byte offsets from the beginning of the file. Screenshots of some dump are not helpful, espcially when the dump tool was not specified, and the key for how to interpret the picture is missing.

**Comment by necros2k7** [comment]
2 screenshots above depict standard header start of windows executable file compressed by UPX and end part of executable file, which can be seen with any file viewer by hitting F3

**Comment by ml1050** [comment]
If you use a proper tool which understands the PE file format then you will see that those "empty" sections are needed there.

**Comment by necros2k7** [comment]
If it`s file format necessity then I agree, discussion can be closed

**Comment by jreiser** [comment]
1. Being specific really helps a discussion.  If a file is involved, then name the file and give a URL.  If a tool is involved, then name the tool and give a URL to it and other similar tools.  If a diagram or picture is involved, then give the key to interpreting.  In this case, does a dot represent any unprintable byte, or some specific byte value?  My "file viewer" is the annotated byte dumper `od`; or the text editor `vi`, which displays non-printable bytes as control characters (NUL is "^@") or hex values such as "<83>", such that different byte values have different representations.
2. An improvement in speed, space, development time or effort, maintenance time or effort, documentation clarity, etc., generally must be at least 1% better than the existing state; otherwise the benefit tends to be not worth the effort.  So reducing the size of a 50KB file by less than 500 bytes, or a 2MB file by less than 20KB, probably won't happen unless there are significant side benefits.

**Comment by necros2k7** [comment]
1. https://github.com/upx/upx/releases/download/v4.0.2/upx-4.0.2-win64.zip (upx.exe)
![1](https://user-images.githubusercontent.com/16452453/224694339-c16576b8-da67-4a24-bfdf-267473b4483b.PNG)
![2](https://user-images.githubusercontent.com/16452453/224694351-34f526ae-372b-4ba3-bbe5-548f2c291b31.PNG)
2. I agree

**Reply by ml1050** [reply]
1. This is the PE header. Not compressible.
2. That's some kind of resource (Side-by-Side Assembly Manifest - no idea what is does... ) In theory it could be compressed, but maybe it was explicitly excluded when the binary was compressed. The reason for not compressed resources is that usually they are accessed by some kind of other program (like the windows explorer).

**Comment by jreiser** [comment]
**necros2k7** second .PNG above, with last line at file offset 0x83230, looks like the information that WindowsExplorer consults when you right click on the desktop icon for an app, then choose Properties.  The last three lines almost certainly are an array of 8-byte structures, with the low 16 bits 0x2000 being some kind of `type` or `flags`, and the high 48 bits being offsets in the file: 0x8226, 0x6426, 0x7226, 0x9026, 0, 0xA026.  (Or, since those 48 bits always end in 0x26, maybe it's something else.)

**Reply by ml1050** [reply]
@jreiser Those last 3 lines are part of the import table.
